{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is my first exploration of the RGI-mapped BN10 genomes to the CARD databases. I used all default settings with RGI, and grabbed only the hits which were deemed \"strict\" (i.e. 100% match). Let's see what's there!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orf_id</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>orientation</th>\n",
       "      <th>cut_off</th>\n",
       "      <th>pass_evalue</th>\n",
       "      <th>best_hit_evalue</th>\n",
       "      <th>best_hit_aro</th>\n",
       "      <th>best_identities</th>\n",
       "      <th>...</th>\n",
       "      <th>pass_bitscore</th>\n",
       "      <th>best_hit_bitscore</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>predicted_dna</th>\n",
       "      <th>predicted_protein</th>\n",
       "      <th>card_protein_sequence</th>\n",
       "      <th>label</th>\n",
       "      <th>id</th>\n",
       "      <th>model_id</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold57|size36017_15</td>\n",
       "      <td>ID=57_15;partial=00;start_type=ATG;rbs_motif=T...</td>\n",
       "      <td>21782</td>\n",
       "      <td>23707</td>\n",
       "      <td>+</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tetQ</td>\n",
       "      <td>96.41</td>\n",
       "      <td>...</td>\n",
       "      <td>1200</td>\n",
       "      <td>1281.540</td>\n",
       "      <td>1281.54</td>\n",
       "      <td>ATGAATATTATAAATTTAGGAATTCTTGCTCACATTGATGCAGGAA...</td>\n",
       "      <td>MNIINLGILAHIDAGKTSVTENLLFASGATEKCGRVDNGDTITDSM...</td>\n",
       "      <td>MRFDNASIVVYYCLIQMNIINLGILAHIDAGKTSVTENLLFASGAT...</td>\n",
       "      <td>scaffold57|size36017_15#21783#23708#1#ID=57_15...</td>\n",
       "      <td>gnl|BL_ORD_ID|2201|hsp_num:0</td>\n",
       "      <td>1183</td>\n",
       "      <td>aa_0143_0005_c5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold238|size8236_1</td>\n",
       "      <td>ID=238_1;partial=00;start_type=ATG;rbs_motif=N...</td>\n",
       "      <td>1445</td>\n",
       "      <td>2245</td>\n",
       "      <td>-</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ErmF</td>\n",
       "      <td>98.50</td>\n",
       "      <td>...</td>\n",
       "      <td>400</td>\n",
       "      <td>531.561</td>\n",
       "      <td>531.561</td>\n",
       "      <td>ATGACAAAAAAGAAATTGCCCGTTCGTTTTACGGGTCAGCACTTTA...</td>\n",
       "      <td>MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...</td>\n",
       "      <td>MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...</td>\n",
       "      <td>scaffold238|size8236_1#1446#2246#-1#ID=238_1;p...</td>\n",
       "      <td>gnl|BL_ORD_ID|159|hsp_num:0</td>\n",
       "      <td>621</td>\n",
       "      <td>aa_0143_0005_c5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold44|size37419_2</td>\n",
       "      <td>ID=44_2;partial=00;start_type=ATG;rbs_motif=No...</td>\n",
       "      <td>1108</td>\n",
       "      <td>2274</td>\n",
       "      <td>-</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tetX</td>\n",
       "      <td>99.74</td>\n",
       "      <td>...</td>\n",
       "      <td>750</td>\n",
       "      <td>799.660</td>\n",
       "      <td>799.66</td>\n",
       "      <td>ATGACAATGCGAATAGATACAGACAAACAAATGAATTTACTTAGTG...</td>\n",
       "      <td>MTMRIDTDKQMNLLSDKNVAIIGGGPVGLTMAKLLQQNGIDVSVYE...</td>\n",
       "      <td>MTMRIDTDKQMNLLSDKNVAIIGGGPVGLTMAKLLQQNGIDVSVYE...</td>\n",
       "      <td>scaffold44|size37419_2#1109#2275#-1#ID=44_2;pa...</td>\n",
       "      <td>gnl|BL_ORD_ID|1398|hsp_num:0</td>\n",
       "      <td>776</td>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold44|size37419_1</td>\n",
       "      <td>ID=44_1;partial=00;start_type=ATG;rbs_motif=No...</td>\n",
       "      <td>95</td>\n",
       "      <td>895</td>\n",
       "      <td>+</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ErmF</td>\n",
       "      <td>99.25</td>\n",
       "      <td>...</td>\n",
       "      <td>400</td>\n",
       "      <td>534.643</td>\n",
       "      <td>534.643</td>\n",
       "      <td>ATGACAAAAAAGAAATTGCCCGTTCGTTTTACGGGTCAGCACTTTA...</td>\n",
       "      <td>MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...</td>\n",
       "      <td>MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...</td>\n",
       "      <td>scaffold44|size37419_1#96#896#1#ID=44_1;partia...</td>\n",
       "      <td>gnl|BL_ORD_ID|159|hsp_num:0</td>\n",
       "      <td>621</td>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold13|size144083_82</td>\n",
       "      <td>ID=13_82;partial=00;start_type=ATG;rbs_motif=N...</td>\n",
       "      <td>110871</td>\n",
       "      <td>112796</td>\n",
       "      <td>+</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tetQ</td>\n",
       "      <td>96.41</td>\n",
       "      <td>...</td>\n",
       "      <td>1200</td>\n",
       "      <td>1281.540</td>\n",
       "      <td>1281.54</td>\n",
       "      <td>ATGAATATTATAAATTTAGGAATTCTTGCTCACATTGATGCAGGAA...</td>\n",
       "      <td>MNIINLGILAHIDAGKTSVTENLLFASGATEKCGRVDNGDTITDSM...</td>\n",
       "      <td>MRFDNASIVVYYCLIQMNIINLGILAHIDAGKTSVTENLLFASGAT...</td>\n",
       "      <td>scaffold13|size144083_82#110872#112797#1#ID=13...</td>\n",
       "      <td>gnl|BL_ORD_ID|2201|hsp_num:0</td>\n",
       "      <td>1183</td>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     orf_id  \\\n",
       "0   scaffold57|size36017_15   \n",
       "1    scaffold238|size8236_1   \n",
       "2    scaffold44|size37419_2   \n",
       "3    scaffold44|size37419_1   \n",
       "4  scaffold13|size144083_82   \n",
       "\n",
       "                                              contig   start    stop  \\\n",
       "0  ID=57_15;partial=00;start_type=ATG;rbs_motif=T...   21782   23707   \n",
       "1  ID=238_1;partial=00;start_type=ATG;rbs_motif=N...    1445    2245   \n",
       "2  ID=44_2;partial=00;start_type=ATG;rbs_motif=No...    1108    2274   \n",
       "3  ID=44_1;partial=00;start_type=ATG;rbs_motif=No...      95     895   \n",
       "4  ID=13_82;partial=00;start_type=ATG;rbs_motif=N...  110871  112796   \n",
       "\n",
       "  orientation cut_off  pass_evalue  best_hit_evalue best_hit_aro  \\\n",
       "0           +  Strict          NaN              0.0         tetQ   \n",
       "1           -  Strict          NaN              0.0         ErmF   \n",
       "2           -  Strict          NaN              0.0         tetX   \n",
       "3           +  Strict          NaN              0.0         ErmF   \n",
       "4           +  Strict          NaN              0.0         tetQ   \n",
       "\n",
       "   best_identities       ...        pass_bitscore best_hit_bitscore bit_score  \\\n",
       "0            96.41       ...                 1200          1281.540   1281.54   \n",
       "1            98.50       ...                  400           531.561   531.561   \n",
       "2            99.74       ...                  750           799.660    799.66   \n",
       "3            99.25       ...                  400           534.643   534.643   \n",
       "4            96.41       ...                 1200          1281.540   1281.54   \n",
       "\n",
       "                                       predicted_dna  \\\n",
       "0  ATGAATATTATAAATTTAGGAATTCTTGCTCACATTGATGCAGGAA...   \n",
       "1  ATGACAAAAAAGAAATTGCCCGTTCGTTTTACGGGTCAGCACTTTA...   \n",
       "2  ATGACAATGCGAATAGATACAGACAAACAAATGAATTTACTTAGTG...   \n",
       "3  ATGACAAAAAAGAAATTGCCCGTTCGTTTTACGGGTCAGCACTTTA...   \n",
       "4  ATGAATATTATAAATTTAGGAATTCTTGCTCACATTGATGCAGGAA...   \n",
       "\n",
       "                                   predicted_protein  \\\n",
       "0  MNIINLGILAHIDAGKTSVTENLLFASGATEKCGRVDNGDTITDSM...   \n",
       "1  MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...   \n",
       "2  MTMRIDTDKQMNLLSDKNVAIIGGGPVGLTMAKLLQQNGIDVSVYE...   \n",
       "3  MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...   \n",
       "4  MNIINLGILAHIDAGKTSVTENLLFASGATEKCGRVDNGDTITDSM...   \n",
       "\n",
       "                               card_protein_sequence  \\\n",
       "0  MRFDNASIVVYYCLIQMNIINLGILAHIDAGKTSVTENLLFASGAT...   \n",
       "1  MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...   \n",
       "2  MTMRIDTDKQMNLLSDKNVAIIGGGPVGLTMAKLLQQNGIDVSVYE...   \n",
       "3  MTKKKLPVRFTGQHFTIDKVLIKDAIRQANISNQDTVLDIGAGKGF...   \n",
       "4  MRFDNASIVVYYCLIQMNIINLGILAHIDAGKTSVTENLLFASGAT...   \n",
       "\n",
       "                                               label  \\\n",
       "0  scaffold57|size36017_15#21783#23708#1#ID=57_15...   \n",
       "1  scaffold238|size8236_1#1446#2246#-1#ID=238_1;p...   \n",
       "2  scaffold44|size37419_2#1109#2275#-1#ID=44_2;pa...   \n",
       "3  scaffold44|size37419_1#96#896#1#ID=44_1;partia...   \n",
       "4  scaffold13|size144083_82#110872#112797#1#ID=13...   \n",
       "\n",
       "                             id model_id           genome  \n",
       "0  gnl|BL_ORD_ID|2201|hsp_num:0     1183  aa_0143_0005_c5  \n",
       "1   gnl|BL_ORD_ID|159|hsp_num:0      621  aa_0143_0005_c5  \n",
       "2  gnl|BL_ORD_ID|1398|hsp_num:0      776  cx_0004_0077_d9  \n",
       "3   gnl|BL_ORD_ID|159|hsp_num:0      621  cx_0004_0077_d9  \n",
       "4  gnl|BL_ORD_ID|2201|hsp_num:0     1183  cx_0004_0077_d9  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fname = '../../data/clean/bn10.rgi_mapped.strict.txt'\n",
    "\n",
    "cols = ['ORF_ID', 'CONTIG', 'START', 'STOP', 'ORIENTATION',\n",
    "        'CUT_OFF', 'PASS_EVALUE', 'Best_Hit_evalue',\n",
    "        'Best_Hit_ARO', 'Best_Identities', 'ARO', 'ARO_name',\n",
    "        'Model_type', 'SNP', 'Best_Hit_ARO_category',\n",
    "        'ARO_category', 'PASS_bitscore', 'Best_Hit_bitscore',\n",
    "        'bit_score', 'Predicted_DNA', 'Predicted_Protein',\n",
    "        'CARD_Protein_Sequence', 'LABEL', 'ID', 'Model_id',\n",
    "        'genome']\n",
    "cols = [c.lower() for c in cols]\n",
    "df = pd.read_csv(fname, sep='\\t', names=cols)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>taxa</th>\n",
       "      <th>donor</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>day</th>\n",
       "      <th>media</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>am_0224_0086_a4</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>am</td>\n",
       "      <td>224</td>\n",
       "      <td>86</td>\n",
       "      <td>a4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>am_0171_0066_c8</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>am</td>\n",
       "      <td>171</td>\n",
       "      <td>66</td>\n",
       "      <td>c8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>am_0171_0071_c9</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>am</td>\n",
       "      <td>171</td>\n",
       "      <td>71</td>\n",
       "      <td>c9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>am_0224_0086_h4</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>am</td>\n",
       "      <td>224</td>\n",
       "      <td>86</td>\n",
       "      <td>h4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>am_0014_0091_a10</td>\n",
       "      <td>Akkermansia_muciniphila</td>\n",
       "      <td>am</td>\n",
       "      <td>14</td>\n",
       "      <td>91</td>\n",
       "      <td>a10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             genome                     taxa donor  genome_id  day media\n",
       "0   am_0224_0086_a4  Akkermansia_muciniphila    am        224   86    a4\n",
       "1   am_0171_0066_c8  Akkermansia_muciniphila    am        171   66    c8\n",
       "2   am_0171_0071_c9  Akkermansia_muciniphila    am        171   71    c9\n",
       "3   am_0224_0086_h4  Akkermansia_muciniphila    am        224   86    h4\n",
       "4  am_0014_0091_a10  Akkermansia_muciniphila    am         14   91   a10"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fmeta = '../../data/raw/table_BN10_WGS_metadata_03282018.txt'\n",
    "# Not sure what hte last four columns are, guess for now\n",
    "cols = ['genome', 'taxa', 'nan', 'donor', 'time_point', 'plate_id', 'well_id']\n",
    "meta = pd.read_csv(fmeta, sep='\\t', names=cols)\n",
    "# One of the columns is all NaN\n",
    "meta = meta.dropna(how='all', axis=1)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "((19315, 26), (3707, 6))\n",
      "(19315, 31)\n"
     ]
    }
   ],
   "source": [
    "# Merge on common genome name\n",
    "print(df.shape, meta.shape)\n",
    "df = pd.merge(df, meta)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19315, 32)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orf_id</th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>orientation</th>\n",
       "      <th>cut_off</th>\n",
       "      <th>pass_evalue</th>\n",
       "      <th>best_hit_evalue</th>\n",
       "      <th>best_hit_aro</th>\n",
       "      <th>best_identities</th>\n",
       "      <th>...</th>\n",
       "      <th>label</th>\n",
       "      <th>id</th>\n",
       "      <th>model_id</th>\n",
       "      <th>genome</th>\n",
       "      <th>taxa</th>\n",
       "      <th>donor</th>\n",
       "      <th>genome_id</th>\n",
       "      <th>day</th>\n",
       "      <th>media</th>\n",
       "      <th>n_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold57|size36017_15</td>\n",
       "      <td>ID=57_15;partial=00;start_type=ATG;rbs_motif=T...</td>\n",
       "      <td>21782</td>\n",
       "      <td>23707</td>\n",
       "      <td>+</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tetQ</td>\n",
       "      <td>96.41</td>\n",
       "      <td>...</td>\n",
       "      <td>scaffold57|size36017_15#21783#23708#1#ID=57_15...</td>\n",
       "      <td>gnl|BL_ORD_ID|2201|hsp_num:0</td>\n",
       "      <td>1183</td>\n",
       "      <td>aa_0143_0005_c5</td>\n",
       "      <td>Bacteroides_ovatus</td>\n",
       "      <td>aa</td>\n",
       "      <td>143</td>\n",
       "      <td>5</td>\n",
       "      <td>c5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold238|size8236_1</td>\n",
       "      <td>ID=238_1;partial=00;start_type=ATG;rbs_motif=N...</td>\n",
       "      <td>1445</td>\n",
       "      <td>2245</td>\n",
       "      <td>-</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ErmF</td>\n",
       "      <td>98.50</td>\n",
       "      <td>...</td>\n",
       "      <td>scaffold238|size8236_1#1446#2246#-1#ID=238_1;p...</td>\n",
       "      <td>gnl|BL_ORD_ID|159|hsp_num:0</td>\n",
       "      <td>621</td>\n",
       "      <td>aa_0143_0005_c5</td>\n",
       "      <td>Bacteroides_ovatus</td>\n",
       "      <td>aa</td>\n",
       "      <td>143</td>\n",
       "      <td>5</td>\n",
       "      <td>c5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold44|size37419_2</td>\n",
       "      <td>ID=44_2;partial=00;start_type=ATG;rbs_motif=No...</td>\n",
       "      <td>1108</td>\n",
       "      <td>2274</td>\n",
       "      <td>-</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tetX</td>\n",
       "      <td>99.74</td>\n",
       "      <td>...</td>\n",
       "      <td>scaffold44|size37419_2#1109#2275#-1#ID=44_2;pa...</td>\n",
       "      <td>gnl|BL_ORD_ID|1398|hsp_num:0</td>\n",
       "      <td>776</td>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "      <td>Bacteroides_xylanisolvens</td>\n",
       "      <td>cx</td>\n",
       "      <td>4</td>\n",
       "      <td>77</td>\n",
       "      <td>d9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold44|size37419_1</td>\n",
       "      <td>ID=44_1;partial=00;start_type=ATG;rbs_motif=No...</td>\n",
       "      <td>95</td>\n",
       "      <td>895</td>\n",
       "      <td>+</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ErmF</td>\n",
       "      <td>99.25</td>\n",
       "      <td>...</td>\n",
       "      <td>scaffold44|size37419_1#96#896#1#ID=44_1;partia...</td>\n",
       "      <td>gnl|BL_ORD_ID|159|hsp_num:0</td>\n",
       "      <td>621</td>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "      <td>Bacteroides_xylanisolvens</td>\n",
       "      <td>cx</td>\n",
       "      <td>4</td>\n",
       "      <td>77</td>\n",
       "      <td>d9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold13|size144083_82</td>\n",
       "      <td>ID=13_82;partial=00;start_type=ATG;rbs_motif=N...</td>\n",
       "      <td>110871</td>\n",
       "      <td>112796</td>\n",
       "      <td>+</td>\n",
       "      <td>Strict</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>tetQ</td>\n",
       "      <td>96.41</td>\n",
       "      <td>...</td>\n",
       "      <td>scaffold13|size144083_82#110872#112797#1#ID=13...</td>\n",
       "      <td>gnl|BL_ORD_ID|2201|hsp_num:0</td>\n",
       "      <td>1183</td>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "      <td>Bacteroides_xylanisolvens</td>\n",
       "      <td>cx</td>\n",
       "      <td>4</td>\n",
       "      <td>77</td>\n",
       "      <td>d9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     orf_id  \\\n",
       "0   scaffold57|size36017_15   \n",
       "1    scaffold238|size8236_1   \n",
       "2    scaffold44|size37419_2   \n",
       "3    scaffold44|size37419_1   \n",
       "4  scaffold13|size144083_82   \n",
       "\n",
       "                                              contig   start    stop  \\\n",
       "0  ID=57_15;partial=00;start_type=ATG;rbs_motif=T...   21782   23707   \n",
       "1  ID=238_1;partial=00;start_type=ATG;rbs_motif=N...    1445    2245   \n",
       "2  ID=44_2;partial=00;start_type=ATG;rbs_motif=No...    1108    2274   \n",
       "3  ID=44_1;partial=00;start_type=ATG;rbs_motif=No...      95     895   \n",
       "4  ID=13_82;partial=00;start_type=ATG;rbs_motif=N...  110871  112796   \n",
       "\n",
       "  orientation cut_off  pass_evalue  best_hit_evalue best_hit_aro  \\\n",
       "0           +  Strict          NaN              0.0         tetQ   \n",
       "1           -  Strict          NaN              0.0         ErmF   \n",
       "2           -  Strict          NaN              0.0         tetX   \n",
       "3           +  Strict          NaN              0.0         ErmF   \n",
       "4           +  Strict          NaN              0.0         tetQ   \n",
       "\n",
       "   best_identities  ...                                                label  \\\n",
       "0            96.41  ...    scaffold57|size36017_15#21783#23708#1#ID=57_15...   \n",
       "1            98.50  ...    scaffold238|size8236_1#1446#2246#-1#ID=238_1;p...   \n",
       "2            99.74  ...    scaffold44|size37419_2#1109#2275#-1#ID=44_2;pa...   \n",
       "3            99.25  ...    scaffold44|size37419_1#96#896#1#ID=44_1;partia...   \n",
       "4            96.41  ...    scaffold13|size144083_82#110872#112797#1#ID=13...   \n",
       "\n",
       "                             id model_id           genome  \\\n",
       "0  gnl|BL_ORD_ID|2201|hsp_num:0     1183  aa_0143_0005_c5   \n",
       "1   gnl|BL_ORD_ID|159|hsp_num:0      621  aa_0143_0005_c5   \n",
       "2  gnl|BL_ORD_ID|1398|hsp_num:0      776  cx_0004_0077_d9   \n",
       "3   gnl|BL_ORD_ID|159|hsp_num:0      621  cx_0004_0077_d9   \n",
       "4  gnl|BL_ORD_ID|2201|hsp_num:0     1183  cx_0004_0077_d9   \n",
       "\n",
       "                        taxa donor  genome_id  day media n_hits  \n",
       "0         Bacteroides_ovatus    aa        143    5    c5      2  \n",
       "1         Bacteroides_ovatus    aa        143    5    c5      2  \n",
       "2  Bacteroides_xylanisolvens    cx          4   77    d9      3  \n",
       "3  Bacteroides_xylanisolvens    cx          4   77    d9      3  \n",
       "4  Bacteroides_xylanisolvens    cx          4   77    d9      3  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count number of hits per genome\n",
    "hits = df.groupby('genome').size().reset_index()\n",
    "hits.columns = ['genome', 'n_hits']\n",
    "\n",
    "resdf = pd.merge(df, hits, how='left')\n",
    "print(resdf.shape)\n",
    "resdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, first I want to just visualize the number of AMR genes per genome.\n",
    "\n",
    "I have the tree from Mathieu's email uploaded into ITOL, so I need to create a gradient dataset annotation: https://itol.embl.de/help/dataset_gradient_template.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "DATASET_GRADIENT\n",
    "#In gradient datasets, each ID is associated to a single numeric value which is converted to a colored box based on the gradient defined.\n",
    "\n",
    "#=================================================================#\n",
    "#                    MANDATORY SETTINGS                           #\n",
    "#=================================================================#\n",
    "#select the separator which is used to delimit the data below (TAB,SPACE or COMMA).This separator must be used throught this file (except in the SEPARATOR line, which uses space).\n",
    "SEPARATOR TAB\n",
    "\n",
    "#label is used in the legend table (can be changed later)\n",
    "DATASET_LABEL n_amr_genes_strict\n",
    "\n",
    "#dataset color (can be changed later)\n",
    "COLOR #ff0000\n",
    "\n",
    "#=================================================================#\n",
    "#                    OPTIONAL SETTINGS                            #\n",
    "#=================================================================#\n",
    "\n",
    "#each dataset can have a legend, which is defined below\n",
    "#for each row in the legend, there should be one shape, color and label\n",
    "#shape should be a number between 1 and 6:\n",
    "#1: square\n",
    "#2: circle\n",
    "#3: star\n",
    "#4: right pointing triangle\n",
    "#5: left pointing triangle\n",
    "#6: checkmark\n",
    "\n",
    "#LEGEND_TITLE Dataset_legend\n",
    "#LEGEND_SHAPES 1 2 3\n",
    "#LEGEND_COLORS #ff0000 #00ff00 #0000ff\n",
    "#LEGEND_LABELS value1 value2 value3\n",
    "\n",
    "#=================================================================#\n",
    "#     all other optional settings can be set or changed later     #\n",
    "#           in the web interface (under 'Datasets' tab)           #\n",
    "#=================================================================#\n",
    "\n",
    "#width of the gradient strip\n",
    "#STRIP_WIDTH 25\n",
    "\n",
    "#left margin, used to increase/decrease the spacing to the next dataset. Can be negative, causing datasets to overlap.\n",
    "#MARGIN 0\n",
    "\n",
    "#border width; if set above 0, a border of specified width (in pixels) will be drawn around the gradient strip\n",
    "#BORDER_WIDTH 0\n",
    "\n",
    "#border color; used whern BORDER_WIDTH is above 0\n",
    "#BORDER_COLOR #0000ff\n",
    "\n",
    "#define the gradient colors. Values in the dataset will be mapped onto the corresponding color gradient.\n",
    "#COLOR_MIN #ff0000\n",
    "#COLOR_MAX #0000ff\n",
    "\n",
    "#you can specify a gradient with three colors (e.g red to yellow to green) by setting 'USE_MID_COLOR' to 1, and specifying the midpoint color\n",
    "#USE_MID_COLOR 1\n",
    "#COLOR_MID #ffff00\n",
    "\n",
    "#always show internal values; if set, values associated to internal nodes will be displayed even if these nodes are not collapsed. It could cause overlapping in the dataset display.\n",
    "#SHOW_INTERNAL 1\n",
    "\n",
    "#Internal tree nodes can be specified using IDs directly, or using the 'last common ancestor' method described in iTOL help pages\n",
    "#=================================================================#\n",
    "#       Actual data follows after the \"DATA\" keyword              #\n",
    "#=================================================================#\n",
    "DATA\n",
    "#ID1 value1\n",
    "#ID2 value2\n",
    "#9606 10000\n",
    "#LEAF1|LEAF2 11000\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open('../../data/analysis/itol_dataset.gradient.n_strict_hits.txt', 'w') as f:\n",
    "    f.write('DATASET_GRADIENT\\n')\n",
    "    f.write('SEPARATOR TAB\\n')\n",
    "    f.write('DATASET_LABEL\\tn_amr_genes_strict\\n')\n",
    "    f.write('COLOR\\t#ff0000\\n')\n",
    "    \n",
    "    # Min = white, max = black\n",
    "    f.write('COLOR_MIN\\t#ffffff\\n')\n",
    "    f.write('COLOR_MAX\\t#000000\\n')\n",
    "    \n",
    "    f.write('DATA\\n')\n",
    "    \n",
    "    for i, j in zip(resdf['genome'], resdf['n_hits']):\n",
    "        f.write(i + '\\t' + str(j) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "taxa                   n_hits\n",
       "Akkermansia_sp.        1           3\n",
       "Alistipes_finegoldii   1           1\n",
       "Alistipes_onderdonkii  1         265\n",
       "                       8           8\n",
       "Alistipes_shahii       1           1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resdf.sort_values(by='n_hits', ascending=False).groupby(['taxa', 'n_hits']).size().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>taxa</th>\n",
       "      <th>n_hits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aa_0143_0005_c5</td>\n",
       "      <td>Bacteroides_ovatus</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aa_0143_0005_c5</td>\n",
       "      <td>Bacteroides_ovatus</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "      <td>Bacteroides_xylanisolvens</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "      <td>Bacteroides_xylanisolvens</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cx_0004_0077_d9</td>\n",
       "      <td>Bacteroides_xylanisolvens</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            genome                       taxa  n_hits\n",
       "0  aa_0143_0005_c5         Bacteroides_ovatus       2\n",
       "1  aa_0143_0005_c5         Bacteroides_ovatus       2\n",
       "2  cx_0004_0077_d9  Bacteroides_xylanisolvens       3\n",
       "3  cx_0004_0077_d9  Bacteroides_xylanisolvens       3\n",
       "4  cx_0004_0077_d9  Bacteroides_xylanisolvens       3"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resdf[['genome', 'taxa', 'n_hits']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Total number of hits per taxa? \n",
    "# Did the tree already (need to color by phylogeny...)\n",
    "\n",
    "# TODO: make a histogram\n",
    "\n",
    "# Variation of number of hits per taxa?\n",
    "# i.e. do most taxa tend to have the same number of ARGs, or do some vary a lot?\n",
    "\n",
    "# Percent of genomes in each taxa that have at least one ARG\n",
    "# This might be the same info as the variation..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
